subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/projectmodule/DefaultProjectDependencyPublicationResolver.java

```diff
@@ -15,76 +15,107 @@
 import org.gradle.api.internal.project.ProjectStateRegistry;
 import org.gradle.execution.ProjectConfigurer;
+import org.gradle.internal.Cast;
+import org.gradle.internal.lazy.Lazy;
 import org.gradle.internal.logging.text.TreeFormatter;
 import org.gradle.internal.service.scopes.Scopes;
 import org.gradle.internal.service.scopes.ServiceScope;
 import org.gradle.util.Path;
 
-import java.util.ArrayList;
+import javax.annotation.Nullable;
+import java.util.Collection;
+import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Iterator;
+import java.util.LinkedHashMap;
 import java.util.LinkedHashSet;
-import java.util.List;
+import java.util.Map;
+import java.util.Objects;
 import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
 
 /**
- * A service that will resolve a ProjectDependency into publication coordinates, to use for publishing.
- * For now is a simple implementation, but at some point could utilise components in the dependency project, usage in the referencing project, etc.
+ * A service that will resolve a project identity path into publication coordinates. If the target variant
+ * of the project is provided, this resolver will attempt to resolve the coordinates for that variant
+ * specifically. Otherwise, it will attempt to resolve the coordinates of the project's root component as
+ * long as the root component does not span across multiple coordinates.
  */
 @ServiceScope(Scopes.Build.class)
 public class DefaultProjectDependencyPublicationResolver implements ProjectDependencyPublicationResolver {
     private final ProjectPublicationRegistry publicationRegistry;
     private final ProjectConfigurer projectConfigurer;
     private final ProjectStateRegistry projects;
 
+    private final VariantCoordinateResolverCache resolverCache = new VariantCoordinateResolverCache();
+
     public DefaultProjectDependencyPublicationResolver(ProjectPublicationRegistry publicationRegistry, ProjectConfigurer projectConfigurer, ProjectStateRegistry projects) {
         this.publicationRegistry = publicationRegistry;
         this.projectConfigurer = projectConfigurer;
         this.projects = projects;
     }
 
     @Override
-    public <T> T resolve(Class<T> coordsType, Path identityPath) {
+    public <T> T resolveComponent(Class<T> coordsType, Path identityPath) {
+        return withResolver(coordsType, identityPath,
+            VariantCoordinateResolver::getComponentCoordinates
+        );
+    }
+
+    @Nullable
+    @Override
+    public <T> T resolveVariant(Class<T> coordsType, Path identityPath, String variantName) {
+        return withResolver(coordsType, identityPath, resolver ->
+            resolver.getVariantCoordinates(variantName)
+        );
+    }
+
+    /**
+     * Execute the action with a resolver for the given project.
+     */
+    private <T> T withResolver(Class<T> coordsType, Path identityPath, Function<VariantCoordinateResolver<T>, T> action) {
```

Comment:
The type system here is not powerful enough to express the code. 

`fromMutableState` accepts a function. It executes it and returns the value returned by the function. 
`withResolver` does something similar. 

For this reason, it is not possible to put proper nullable annotations on these methods. 

`withResolver` is nullable if the function passed as a parameter is nullable. 
It is non-nullable if the function is non-nullable. 


Since `getComponentCoordinates` is non-nullable, `resolveComponent` is non-nullable. 
Since `getVariantCoordinates` is nullable, `resolveVariant` is nullable.

Reply:
I renamed a couple things to include `CoordinateResolver`